Introduce communty history archive routine

0x-r4bbit opened the pull request:
This introduces logic needed to:

1. Create `WakuMessageArchive`s and and indices from store waku messages
2. History archive torrent data to disk and create `.torrent` file from
   that
3. Seed and unseed history archive torrents as necessary
4. Starting/stopping the torrent client
5. Enabling/disabling community history support for individual
   components and starting/stopping the routine intervals accordingly

This **does not** yet handle magnet links (#2568)

Closes #2567

ghost commented:
<!--prchecklist--> 
### Pull Request Checklist
- [ ] Have you updated the documentation, if impacted (e.g. [docs.status.im](https://docs.status.im/docs/build_status_go.html))?
- [ ] Have you tested changes with mobile?
- [ ] Have you tested changes with desktop?

0x-r4bbit commented:
This still needs tests.

ghost commented:
`vendor/github.com/pion/webrtc/v3/yarn.lock` changed. Pinging @adambabik and @corpetty

ghost commented:
`vendor/github.com/pion/webrtc/v3/yarn.lock` changed. Pinging @adambabik and @corpetty

0x-r4bbit commented on protocol/communities/manager.go:
```diff
@@ -15,23 +20,36 @@ import (
+	{"udp://tracker.opentrackr.org:1337/announce"},
+	{"udp://tracker.openbittorrent.com:6969/announce"},
+}
+var pieceLength = 1024
```
This value is for testing purposes and needs to be updated

status-im-auto commented:
### Jenkins Builds
<details>
<summary>Click to see older builds (60)</summary>

| :grey_question: | Commit | :hash: | Finished (UTC) | Duration | Platform | Result |
|-|-|-|-|-|-|-|
| :heavy_check_mark: | 24570486 | [#1](https://ci.status.im/job/status-go/job/prs/job/linux/job/PR-2576/1/) | 2022-03-09 13:03:18 | ~3 min | `linux` | [:package:`zip`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/status-go-desktop-220309-125947-245704-pr2576.zip) |
| :heavy_check_mark: | 24570486 | [#1](https://ci.status.im/job/status-go/job/prs/job/android/job/PR-2576/1/) | 2022-03-09 13:04:40 | ~4 min | `android` | [:package:`aar`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/status-go-android-220309-125943-245704-pr2576.aar) |
| :heavy_check_mark: | 24570486 | [#1](https://ci.status.im/job/status-go/job/prs/job/ios/job/PR-2576/1/) | 2022-03-09 13:05:46 | ~5 min | `ios` | [:package:`zip`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/status-go-ios-220309-125947-245704-pr2576.zip) |
| | | | | | | |
| :heavy_check_mark: | 94b3d60b | [#2](https://ci.status.im/job/status-go/job/prs/job/ios/job/PR-2576/2/) | 2022-03-10 09:25:58 | ~2 min | `ios` | [:package:`zip`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/status-go-ios-220310-092325-94b3d6-pr2576.zip) |
| :heavy_check_mark: | 94b3d60b | [#2](https://ci.status.im/job/status-go/job/prs/job/linux/job/PR-2576/2/) | 2022-03-10 09:26:59 | ~3 min | `linux` | [:package:`zip`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/status-go-desktop-220310-092325-94b3d6-pr2576.zip) |
| :heavy_check_mark: | 94b3d60b | [#2](https://ci.status.im/job/status-go/job/prs/job/android/job/PR-2576/2/) | 2022-03-10 09:27:51 | ~4 min | `android` | [:package:`aar`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/status-go-android-220310-092325-94b3d6-pr2576.aar) |
| :heavy_check_mark: | 94b3d60b | [#3](https://ci.status.im/job/status-go/job/prs/job/ios/job/PR-2576/3/) | 2022-03-10 09:27:59 | ~1 min | `ios` | [:package:`zip`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/status-go-ios-220310-092603-94b3d6-pr2576.zip) |
| :heavy_check_mark: | 94b3d60b | [#3](https://ci.status.im/job/status-go/job/prs/job/linux/job/PR-2576/3/) | 2022-03-10 09:28:32 | ~1 min | `linux` | [:package:`zip`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/status-go-desktop-220310-092701-94b3d6-pr2576.zip) |
| :heavy_check_mark: | 94b3d60b | [#3](https://ci.status.im/job/status-go/job/prs/job/android/job/PR-2576/3/) | 2022-03-10 09:31:07 | ~3 min | `android` | [:package:`aar`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/status-go-android-220310-092753-94b3d6-pr2576.aar) |
| | | | | | | |
| :heavy_check_mark: | 61e685af | [#4](https://ci.status.im/job/status-go/job/prs/job/linux/job/PR-2576/4/) | 2022-03-10 09:31:28 | ~1 min | `linux` | [:package:`zip`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/status-go-desktop-220310-092935-61e685-pr2576.zip) |
| :heavy_check_mark: | 61e685af | [#4](https://ci.status.im/job/status-go/job/prs/job/ios/job/PR-2576/4/) | 2022-03-10 09:32:12 | ~2 min | `ios` | [:package:`zip`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/status-go-ios-220310-092935-61e685-pr2576.zip) |
| :heavy_check_mark: | 61e685af | [#4](https://ci.status.im/job/status-go/job/prs/job/android/job/PR-2576/4/) | 2022-03-10 09:36:28 | ~5 min | `android` | [:package:`aar`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/status-go-android-220310-093112-61e685-pr2576.aar) |
| | | | | | | |
| :heavy_check_mark: | 3e6dc7ce | [#5](https://ci.status.im/job/status-go/job/prs/job/linux/job/PR-2576/5/) | 2022-03-10 09:53:00 | ~2 min | `linux` | [:package:`zip`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/status-go-desktop-220310-095031-3e6dc7-pr2576.zip) |
| :heavy_check_mark: | 3e6dc7ce | [#5](https://ci.status.im/job/status-go/job/prs/job/ios/job/PR-2576/5/) | 2022-03-10 09:54:35 | ~4 min | `ios` | [:package:`zip`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/status-go-ios-220310-095031-3e6dc7-pr2576.zip) |
| :heavy_check_mark: | 3e6dc7ce | [#5](https://ci.status.im/job/status-go/job/prs/job/android/job/PR-2576/5/) | 2022-03-10 09:54:47 | ~4 min | `android` | [:package:`aar`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/status-go-android-220310-095031-3e6dc7-pr2576.aar) |
| | | | | | | |
| :heavy_check_mark: | 10a51def | [#6](https://ci.status.im/job/status-go/job/prs/job/linux/job/PR-2576/6/) | 2022-03-14 12:03:52 | ~2 min | `linux` | [:package:`zip`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/status-go-desktop-220314-120139-10a51d-pr2576.zip) |
| :heavy_check_mark: | 10a51def | [#6](https://ci.status.im/job/status-go/job/prs/job/ios/job/PR-2576/6/) | 2022-03-14 12:04:31 | ~2 min | `ios` | [:package:`zip`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/status-go-ios-220314-120139-10a51d-pr2576.zip) |
| :heavy_check_mark: | 10a51def | [#6](https://ci.status.im/job/status-go/job/prs/job/android/job/PR-2576/6/) | 2022-03-14 12:05:20 | ~3 min | `android` | [:package:`aar`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/status-go-android-220314-120139-10a51d-pr2576.aar) |
| | | | | | | |
| :heavy_check_mark: | 5881a6dc | [#7](https://ci.status.im/job/status-go/job/prs/job/linux/job/PR-2576/7/) | 2022-03-21 14:50:13 | ~2 min | `linux` | [:package:`zip`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/status-go-desktop-220321-144734-5881a6-pr2576.zip) |
| :heavy_check_mark: | 5881a6dc | [#7](https://ci.status.im/job/status-go/job/prs/job/ios/job/PR-2576/7/) | 2022-03-21 14:51:39 | ~4 min | `ios` | [:package:`zip`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/status-go-ios-220321-144734-5881a6-pr2576.zip) |
| :heavy_check_mark: | 5881a6dc | [#7](https://ci.status.im/job/status-go/job/prs/job/android/job/PR-2576/7/) | 2022-03-21 14:51:52 | ~4 min | `android` | [:package:`aar`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/status-go-android-220321-144730-5881a6-pr2576.aar) |
| | | | | | | |
| :heavy_check_mark: | 3e1eb0a2 | [#8](https://ci.status.im/job/status-go/job/prs/job/linux/job/PR-2576/8/) | 2022-03-21 14:55:52 | ~2 min | `linux` | [:package:`zip`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/status-go-desktop-220321-145328-3e1eb0-pr2576.zip) |
| :heavy_check_mark: | 3e1eb0a2 | [#8](https://ci.status.im/job/status-go/job/prs/job/ios/job/PR-2576/8/) | 2022-03-21 14:56:53 | ~3 min | `ios` | [:package:`zip`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/status-go-ios-220321-145328-3e1eb0-pr2576.zip) |
| :heavy_check_mark: | 3e1eb0a2 | [#8](https://ci.status.im/job/status-go/job/prs/job/android/job/PR-2576/8/) | 2022-03-21 14:57:13 | ~3 min | `android` | [:package:`aar`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/status-go-android-220321-145325-3e1eb0-pr2576.aar) |
| | | | | | | |
| :heavy_check_mark: | b10fe618 | [#9](https://ci.status.im/job/status-go/job/prs/job/linux/job/PR-2576/9/) | 2022-03-22 14:14:58 | ~3 min | `linux` | [:package:`zip`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/status-go-desktop-220322-141145-b10fe6-pr2576.zip) |
| :heavy_check_mark: | b10fe618 | [#9](https://ci.status.im/job/status-go/job/prs/job/android/job/PR-2576/9/) | 2022-03-22 14:17:13 | ~5 min | `android` | [:package:`aar`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/status-go-android-220322-141145-b10fe6-pr2576.aar) |
| :heavy_check_mark: | b10fe618 | [#9](https://ci.status.im/job/status-go/job/prs/job/ios/job/PR-2576/9/) | 2022-03-22 14:17:32 | ~5 min | `ios` | [:package:`zip`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/status-go-ios-220322-141145-b10fe6-pr2576.zip) |
| | | | | | | |
| :heavy_check_mark: | df405e2d | [#10](https://ci.status.im/job/status-go/job/prs/job/linux/job/PR-2576/10/) | 2022-03-22 16:26:47 | ~1 min | `linux` | [:package:`zip`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/status-go-desktop-220322-162511-df405e-pr2576.zip) |
| :heavy_check_mark: | df405e2d | [#10](https://ci.status.im/job/status-go/job/prs/job/android/job/PR-2576/10/) | 2022-03-22 16:28:31 | ~3 min | `android` | [:package:`aar`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/status-go-android-220322-162506-df405e-pr2576.aar) |
| :heavy_check_mark: | df405e2d | [#10](https://ci.status.im/job/status-go/job/prs/job/ios/job/PR-2576/10/) | 2022-03-22 16:32:08 | ~6 min | `ios` | [:package:`zip`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/status-go-ios-220322-162511-df405e-pr2576.zip) |
| | | | | | | |
| :heavy_check_mark: | 2bf79e94 | [#11](https://ci.status.im/job/status-go/job/prs/job/linux/job/PR-2576/11/) | 2022-03-23 18:26:42 | ~1 min | `linux` | [:package:`zip`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/status-go-desktop-220323-182505-2bf79e-pr2576.zip) |
| :heavy_check_mark: | 2bf79e94 | [#11](https://ci.status.im/job/status-go/job/prs/job/ios/job/PR-2576/11/) | 2022-03-23 18:28:00 | ~2 min | `ios` | [:package:`zip`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/status-go-ios-220323-182505-2bf79e-pr2576.zip) |
| :heavy_check_mark: | 2bf79e94 | [#11](https://ci.status.im/job/status-go/job/prs/job/android/job/PR-2576/11/) | 2022-03-23 18:28:33 | ~3 min | `android` | [:package:`aar`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/status-go-android-220323-182503-2bf79e-pr2576.aar) |
| | | | | | | |
| :heavy_check_mark: | 0f97c612 | [#12](https://ci.status.im/job/status-go/job/prs/job/ios/job/PR-2576/12/) | 2022-03-25 11:26:58 | ~5 min | `ios` | [:package:`zip`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/status-go-ios-220325-112102-0f97c6-pr2576.zip) |
| :heavy_check_mark: | 0f97c612 | [#12](https://ci.status.im/job/status-go/job/prs/job/linux/job/PR-2576/12/) | 2022-03-25 11:29:39 | ~8 min | `linux` | [:package:`zip`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/status-go-desktop-220325-112107-0f97c6-pr2576.zip) |
| :heavy_check_mark: | 0f97c612 | [#12](https://ci.status.im/job/status-go/job/prs/job/android/job/PR-2576/12/) | 2022-03-25 11:30:14 | ~9 min | `android` | [:package:`aar`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/status-go-android-220325-112100-0f97c6-pr2576.aar) |
| | | | | | | |
| :heavy_check_mark: | 835dffa2 | [#13](https://ci.status.im/job/status-go/job/prs/job/ios/job/PR-2576/13/) | 2022-03-25 14:22:00 | ~2 min | `ios` | [:package:`zip`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/status-go-ios-220325-141935-835dff-pr2576.zip) |
| :heavy_check_mark: | 835dffa2 | [#13](https://ci.status.im/job/status-go/job/prs/job/linux/job/PR-2576/13/) | 2022-03-25 14:23:00 | ~3 min | `linux` | [:package:`zip`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/status-go-desktop-220325-141935-835dff-pr2576.zip) |
| :heavy_check_mark: | 835dffa2 | [#13](https://ci.status.im/job/status-go/job/prs/job/android/job/PR-2576/13/) | 2022-03-25 14:24:48 | ~5 min | `android` | [:package:`aar`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/status-go-android-220325-141935-835dff-pr2576.aar) |
| | | | | | | |
| :heavy_check_mark: | 7ce4d90c | [#14](https://ci.status.im/job/status-go/job/prs/job/linux/job/PR-2576/14/) | 2022-03-28 15:53:28 | ~2 min | `linux` | [:package:`zip`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/status-go-desktop-220328-155055-7ce4d9-pr2576.zip) |
| :heavy_check_mark: | 7ce4d90c | [#14](https://ci.status.im/job/status-go/job/prs/job/ios/job/PR-2576/14/) | 2022-03-28 15:54:13 | ~3 min | `ios` | [:package:`zip`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/status-go-ios-220328-155055-7ce4d9-pr2576.zip) |
| :heavy_check_mark: | 7ce4d90c | [#14](https://ci.status.im/job/status-go/job/prs/job/android/job/PR-2576/14/) | 2022-03-28 15:55:34 | ~4 min | `android` | [:package:`aar`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/status-go-android-220328-155055-7ce4d9-pr2576.aar) |
| | | | | | | |
| :heavy_check_mark: | aee513f7 | [#15](https://ci.status.im/job/status-go/job/prs/job/linux/job/PR-2576/15/) | 2022-03-28 16:22:15 | ~4 min | `linux` | [:package:`zip`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/status-go-desktop-220328-161723-aee513-pr2576.zip) |
| :heavy_check_mark: | aee513f7 | [#15](https://ci.status.im/job/status-go/job/prs/job/ios/job/PR-2576/15/) | 2022-03-28 16:23:26 | ~6 min | `ios` | [:package:`zip`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/status-go-ios-220328-161723-aee513-pr2576.zip) |
| :heavy_check_mark: | aee513f7 | [#15](https://ci.status.im/job/status-go/job/prs/job/android/job/PR-2576/15/) | 2022-03-28 16:25:01 | ~7 min | `android` | [:package:`aar`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/status-go-android-220328-161722-aee513-pr2576.aar) |
| | | | | | | |
| :heavy_check_mark: | 8b201d47 | [#16](https://ci.status.im/job/status-go/job/prs/job/linux/job/PR-2576/16/) | 2022-03-30 06:55:39 | ~3 min | `linux` | [:package:`zip`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/status-go-desktop-220330-065200-8b201d-pr2576.zip) |
| :heavy_check_mark: | 8b201d47 | [#16](https://ci.status.im/job/status-go/job/prs/job/ios/job/PR-2576/16/) | 2022-03-30 06:55:49 | ~3 min | `ios` | [:package:`zip`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/status-go-ios-220330-065200-8b201d-pr2576.zip) |
| :heavy_check_mark: | 8b201d47 | [#16](https://ci.status.im/job/status-go/job/prs/job/android/job/PR-2576/16/) | 2022-03-30 06:56:33 | ~4 min | `android` | [:package:`aar`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/status-go-android-220330-065200-8b201d-pr2576.aar) |
| | | | | | | |
| :heavy_check_mark: | cffda3ff | [#17](https://ci.status.im/job/status-go/job/prs/job/ios/job/PR-2576/17/) | 2022-03-30 07:44:06 | ~2 min | `ios` | [:package:`zip`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/status-go-ios-220330-074134-cffda3-pr2576.zip) |
| :heavy_check_mark: | cffda3ff | [#17](https://ci.status.im/job/status-go/job/prs/job/linux/job/PR-2576/17/) | 2022-03-30 07:44:15 | ~2 min | `linux` | [:package:`zip`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/status-go-desktop-220330-074136-cffda3-pr2576.zip) |
| :heavy_check_mark: | cffda3ff | [#17](https://ci.status.im/job/status-go/job/prs/job/android/job/PR-2576/17/) | 2022-03-30 07:44:47 | ~3 min | `android` | [:package:`aar`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/status-go-android-220330-074134-cffda3-pr2576.aar) |
| | | | | | | |
| :heavy_check_mark: | 63d04b98 | [#18](https://ci.status.im/job/status-go/job/prs/job/linux/job/PR-2576/18/) | 2022-04-05 11:19:26 | ~2 min | `linux` | [:package:`zip`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/status-go-desktop-220405-111700-63d04b-pr2576.zip) |
| :heavy_check_mark: | 63d04b98 | [#18](https://ci.status.im/job/status-go/job/prs/job/android/job/PR-2576/18/) | 2022-04-05 11:21:00 | ~4 min | `android` | [:package:`aar`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/status-go-android-220405-111656-63d04b-pr2576.aar) |
| :heavy_check_mark: | 63d04b98 | [#18](https://ci.status.im/job/status-go/job/prs/job/ios/job/PR-2576/18/) | 2022-04-05 11:21:04 | ~4 min | `ios` | [:package:`zip`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/status-go-ios-220405-111700-63d04b-pr2576.zip) |
| | | | | | | |
| :heavy_check_mark: | 0beba92e | [#19](https://ci.status.im/job/status-go/job/prs/job/linux/job/PR-2576/19/) | 2022-04-05 13:21:36 | ~1 min | `linux` | [:package:`zip`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/status-go-desktop-220405-132021-0beba9-pr2576.zip) |
| :heavy_check_mark: | 0beba92e | [#19](https://ci.status.im/job/status-go/job/prs/job/ios/job/PR-2576/19/) | 2022-04-05 13:22:04 | ~1 min | `ios` | [:package:`zip`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/status-go-ios-220405-132021-0beba9-pr2576.zip) |
| :heavy_check_mark: | 0beba92e | [#19](https://ci.status.im/job/status-go/job/prs/job/android/job/PR-2576/19/) | 2022-04-05 13:23:22 | ~3 min | `android` | [:package:`aar`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/status-go-android-220405-132018-0beba9-pr2576.aar) |
| | | | | | | |
| :heavy_check_mark: | eecb4af8 | [#20](https://ci.status.im/job/status-go/job/prs/job/linux/job/PR-2576/20/) | 2022-04-05 13:48:09 | ~1 min | `linux` | [:package:`zip`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/status-go-desktop-220405-134622-eecb4a-pr2576.zip) |
| :heavy_check_mark: | eecb4af8 | [#20](https://ci.status.im/job/status-go/job/prs/job/android/job/PR-2576/20/) | 2022-04-05 13:50:35 | ~4 min | `android` | [:package:`aar`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/status-go-android-220405-134619-eecb4a-pr2576.aar) |
| :heavy_check_mark: | eecb4af8 | [#20](https://ci.status.im/job/status-go/job/prs/job/ios/job/PR-2576/20/) | 2022-04-05 13:51:00 | ~4 min | `ios` | [:package:`zip`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/status-go-ios-220405-134622-eecb4a-pr2576.zip) |
</details>

| :grey_question: | Commit | :hash: | Finished (UTC) | Duration | Platform | Result |
|-|-|-|-|-|-|-|
| :heavy_check_mark: | b0880c94 | [#21](https://ci.status.im/job/status-go/job/prs/job/linux/job/PR-2576/21/) | 2022-04-05 16:52:33 | ~1 min | `linux` | [:package:`zip`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/status-go-desktop-220405-165101-b0880c-pr2576.zip) |
| :heavy_check_mark: | b0880c94 | [#21](https://ci.status.im/job/status-go/job/prs/job/ios/job/PR-2576/21/) | 2022-04-05 16:52:44 | ~1 min | `ios` | [:package:`zip`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/status-go-ios-220405-165101-b0880c-pr2576.zip) |
| :heavy_check_mark: | b0880c94 | [#21](https://ci.status.im/job/status-go/job/prs/job/android/job/PR-2576/21/) | 2022-04-05 16:54:19 | ~3 min | `android` | [:package:`aar`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/status-go-android-220405-165058-b0880c-pr2576.aar) |
| :heavy_check_mark: | b0880c94 | [#22](https://ci.status.im/job/status-go/job/prs/job/linux/job/PR-2576/22/) | 2022-04-06 09:50:03 | ~1 min | `linux` | [:package:`zip`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/status-go-desktop-220406-094836-b0880c-pr2576.zip) |
| :heavy_check_mark: | b0880c94 | [#22](https://ci.status.im/job/status-go/job/prs/job/ios/job/PR-2576/22/) | 2022-04-06 09:50:52 | ~2 min | `ios` | [:package:`zip`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/status-go-ios-220406-094834-b0880c-pr2576.zip) |
| :heavy_check_mark: | b0880c94 | [#22](https://ci.status.im/job/status-go/job/prs/job/android/job/PR-2576/22/) | 2022-04-06 09:51:57 | ~3 min | `android` | [:package:`aar`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/status-go-android-220406-094834-b0880c-pr2576.aar) |
| | | | | | | |
| :heavy_check_mark: | 8e703337 | [#23](https://ci.status.im/job/status-go/job/prs/job/linux/job/PR-2576/23/) | 2022-04-06 10:43:41 | ~1 min | `linux` | [:package:`zip`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/status-go-desktop-220406-104158-8e7033-pr2576.zip) |
| :heavy_check_mark: | 8e703337 | [#23](https://ci.status.im/job/status-go/job/prs/job/ios/job/PR-2576/23/) | 2022-04-06 10:43:51 | ~1 min | `ios` | [:package:`zip`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/status-go-ios-220406-104158-8e7033-pr2576.zip) |
| :heavy_check_mark: | 8e703337 | [#23](https://ci.status.im/job/status-go/job/prs/job/android/job/PR-2576/23/) | 2022-04-06 10:45:00 | ~3 min | `android` | [:package:`aar`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/status-go-android-220406-104157-8e7033-pr2576.aar) |

ghost commented:
`vendor/github.com/pion/webrtc/v3/yarn.lock` changed. Pinging @adambabik and @corpetty

ghost commented:
`vendor/github.com/pion/webrtc/v3/yarn.lock` changed. Pinging @adambabik and @corpetty

ghost commented:
`vendor/github.com/pion/webrtc/v3/yarn.lock` changed. Pinging @adambabik and @corpetty

ghost commented:
`vendor/github.com/pion/webrtc/v3/yarn.lock` changed. Pinging @adambabik and @corpetty

0x-r4bbit commented on protocol/messenger.go:
```diff
@@ -151,11 +151,12 @@ type peerStatus struct {
+	peers                     map[string]peerStatus
+	events                    chan *p2p.PeerEvent
+	subscription              event.Subscription
+	availabilitySubscriptions map[chan struct{}]chan string
```
This was my attempt to create a way to subscribe inside status-go to whether a mailserver is available.

There's probably a better way to do this, please let me know.

0x-r4bbit commented on protocol/messenger.go:
```diff
@@ -659,6 +661,18 @@ func (m *Messenger) Start() (*MessengerResponse, error) {
+				go func() {
+					<-mailserverAvailable
+					close(unsubscribe)
+					m.InitHistoryArchiveTasks(adminCommunities)
```
Here you can see how I wait for `mailserverAvailable` being emitted before we kick off the history archive tasks.

Again, there's probably a nicer more "go-like" way to do this.

0x-r4bbit commented on protocol/messenger_communities.go:
```diff
@@ -139,7 +142,7 @@ func (m *Messenger) JoinCommunity(ctx context.Context, communityID types.HexByte
 
 	communitySettings := communities.CommunitySettings{
 		CommunityID:                  communityID.String(),
-		HistoryArchiveSupportEnabled: true,
```
Ah, this should've probably been `false` for now in the previous commit as well.

0x-r4bbit commented on protocol/messenger_communities.go:
```diff
@@ -1106,6 +1137,220 @@ func (m *Messenger) handleSyncCommunity(messageState *ReceivedMessageState, sync
+			settings, err := m.communitiesManager.GetCommunitySettingsByID(c.ID())
+			if err != nil {
+				log.Println("failed to get community settings", err)
+				m.logger.Debug("failed to get community settings", zap.Error(err))
```
I've doubled the logs here so I can see something in stdout. Happy to remove them for final version, but it's very helpful to see what's going on in realtime..

0x-r4bbit commented on protocol/messenger_communities.go:
```diff
@@ -18,6 +19,8 @@ import (
 	"github.com/status-im/status-go/protocol/transport"
 )
 
+var messageArchiveInterval = 5 * time.Minute
```
This should be 7 * time.Day once this goes to production, but for testing a shorter interval is needed

0x-r4bbit commented on protocol/messenger_mailserver.go:
```diff
@@ -263,7 +263,7 @@ func getPrioritizedBatches() []int {
 }
 
-func (m *Messenger) syncFilters(filters []*transport.Filter) (*MessengerResponse, error) {
+func (m *Messenger) syncFiltersFrom(filters []*transport.Filter, lastRequest uint32) (*MessengerResponse, error) {
```
This is my attempt to have `syncFilters` provide a way to control how far we want to go back in time to sync from. 

If this is not a good way to do this, please let me know. Basically we need a way to say "Fetch all messages from store node from last 30 days"

0x-r4bbit commented on protocol/messenger_mailserver.go:
```diff
@@ -439,6 +448,10 @@ func (m *Messenger) syncFilters(filters []*transport.Filter) (*MessengerResponse
 
+func (m *Messenger) syncFilters(filters []*transport.Filter) (*MessengerResponse, error) {
+	return m.syncFiltersFrom(filters, 0)
+}
```
Notice here how the original `syncFilters` semantics stay intact

0x-r4bbit commented on protocol/messenger_mailserver_cycle.go:
```diff
@@ -648,6 +651,23 @@ func (m *Messenger) checkMailserverConnection() {
+	cancel := make(chan struct{})
+	m.mailserverCycle.availabilitySubscriptions[cancel] = c
+	return c, cancel
+}
```
This is the other part of me trying to get a signal inside status-go that a mailserver is available.

Most likely better ways to do this. Appreciate input

0x-r4bbit commented on protocol/communities/manager.go:
```diff
@@ -957,6 +1045,490 @@ func (m *Manager) GetAdminCommunitiesChatIDs() (map[string]bool, error) {
+			return err
+		}
+
+		wakuMessageArchive := m.createWakuMessageArchive(from, to, messages, topicsAsByteArrays)
```
Don't create archive if no messages for this time range exist

0x-r4bbit replied:
This is now addressed above in L:1284

0x-r4bbit commented on protocol/communities/manager.go:
```diff
@@ -957,6 +1045,490 @@ func (m *Manager) GetAdminCommunitiesChatIDs() (map[string]bool, error) {
+		return err
+	}
+
+	err = m.persistence.UpdateLastMessageArchiveEndDate(communityID, uint64(from.Unix()))
```
This has a bug. We need to check whether we already have one to update it. Otherwise add it.

0x-r4bbit replied:
This has a bug. We need to check whether we already have one to update it. Otherwise add it.

0x-r4bbit committed:
Introduce community history archive routine

This introduces logic needed to:

- Create WakuMessageArchives and and indices from store waku messages
- History archive torrent data to disk and create .torrent file from
  that
- Seed and unseed history archive torrents as necessary
- Starting/stopping the torrent client
- Enabling/disabling community history support for individual components
  and starting/stopping the routine intervals accordingly

This does not yet handle magnet links (#2568)

Closes #2567

0x-r4bbit commented on protocol/communities/manager.go:
```diff
@@ -18,23 +21,37 @@ import (
+	{"udp://tracker.opentrackr.org:1337/announce"},
+	{"udp://tracker.openbittorrent.com:6969/announce"},
+}
+var pieceLength = 100 * 1024
```
Going with 100kb pieces now but we probably want to change this @staheri14 I believe you wanted to propose a good piece size?

0x-r4bbit commented on protocol/messenger_communities.go:
```diff
@@ -18,6 +19,9 @@ import (
 )
 
+// 7 days interval
+var messageArchiveInterval = 7 * 24 * time.Hour
```
For testing purposes, it might make sense to change this for now or at least make it configurable.

0x-r4bbit commented on protocol/messenger_communities.go:
```diff
@@ -1098,6 +1130,226 @@ func (m *Messenger) handleSyncCommunity(messageState *ReceivedMessageState, sync
+				log.Println("no filters or chats for this community", c.IDString())
+				m.logger.Debug("no filters or chats for this community")
+				continue
+			}
```
Actually, this is an issue... When we create a community it won't have any filters because it won't have any chats yet. This means, when `InitHistoryArchiveTasks` is called, it won't schedule the interval. We should schedule it here.

0x-r4bbit replied:
This has been addressed by now..

0x-r4bbit commented on protocol/messenger_mailserver_cycle.go:
```diff
@@ -650,3 +651,18 @@ func (m *Messenger) getPinnedMailserver() (*mailservers.Mailserver, error) {
+	c := make(chan struct{})
+	m.mailserverCycle.availabilitySubscriptions = append(m.mailserverCycle.availabilitySubscriptions, c)
+	return c
+}
```
@Samyoul I've updated this as per our discussions. `EmitMailserverAvailable` now merely emits a signal and closes directly without waiting for a cancel signal from a second channel

0x-r4bbit commented on protocol/communities/manager.go:
```diff
@@ -961,6 +1043,512 @@ func (m *Manager) GetAdminCommunitiesChatIDs() (map[string]bool, error) {
+
+	cancel := make(chan struct{})
+	m.historyArchiveTasks[id] = cancel
+	m.historyArchiveTasksWaitGroup.Add(1)
```
@Samyoul Here I'm making use of `waitGroup`s. Lemme know if this should be done differently.

0x-r4bbit commented on protocol/communities/manager.go:
```diff
@@ -961,6 +1043,512 @@ func (m *Manager) GetAdminCommunitiesChatIDs() (map[string]bool, error) {
+		case <-cancel:
+			m.UnseedHistoryArchiveTorrent(community.ID())
+			delete(m.historyArchiveTasks, id)
+			m.historyArchiveTasksWaitGroup.Done()
```
@Samyoul We're decreasing the waitgroup counter here

0x-r4bbit commented on protocol/communities/manager.go:
```diff
@@ -961,6 +1043,512 @@ func (m *Manager) GetAdminCommunitiesChatIDs() (map[string]bool, error) {
+	// Stoping archive interval tasks is async, so we need
+	// to wait for all of them to be closed before we shutdown
+	// the torrent client
+	m.historyArchiveTasksWaitGroup.Wait()
```
@Samyoul And wait for it to be done here, making `StopHistoryArchiveTasksIntervals` a somewhat synchronous API

0x-r4bbit commented on protocol/messenger.go:
```diff
@@ -623,6 +625,7 @@ func (m *Messenger) Start() (*MessengerResponse, error) {
 
 	m.handleEncryptionLayerSubscriptions(subscriptions)
 	m.handleCommunitiesSubscription(m.communitiesManager.Subscribe())
+	m.handleCommunitiesHistoryArchivesSubscription(m.communitiesManager.Subscribe())
```
I'm reusing the existing community manager subscriptions to signal community history events as well

0x-r4bbit commented on protocol/messenger_communities.go:
```diff
@@ -59,6 +63,56 @@ func (m *Messenger) publishOrgInvitation(org *communities.Community, invitation
+						sub.NoHistoryArchivesCreatedSignal.From,
+						sub.NoHistoryArchivesCreatedSignal.To,
+					)
+				}
```
These look a bit unfortunate. We emit the `History*Signal`'s on the `Subscription` and then use the attached data to send them to connected clients by destructing the attached data, which is then reconstructed again.

The reason I do this is because we already need the `History*Signal`s anyways to inform connected clients, so it made sense to reuse them here in the subscription.

Lemme know if this should be done differently.

0x-r4bbit commented on protocol/communities/manager.go:
```diff
@@ -67,8 +89,13 @@ func NewManager(identity *ecdsa.PublicKey, db *sql.DB, logger *zap.Logger, verif
+	HistoryArchivesCreatedSignal   *signal.HistoryArchivesCreatedSignal
+	NoHistoryArchivesCreatedSignal *signal.NoHistoryArchivesCreatedSignal
+	HistoryArchivesSeedingSignal   *signal.HistoryArchivesSeedingSignal
+	HistoryArchivesUnseededSignal  *signal.HistoryArchivesUnseededSignal
```
I could build a separate `Subscription` type for this, but I thought there's no reason not to put those events in here.

0x-r4bbit commented:
I know this is a large PR, but would be cool if we could get it reviewed @cammellos @Samyoul 

It's been through a bunch of iterations (and have to go through another one as there were changes in `develop` that force a rebase here, also because of the linter needing an upgrade), if you need any info or a walk-through of this, please let me know

0x-r4bbit commented on protocol/messenger_communities.go:
```diff
@@ -1095,6 +1172,223 @@ func (m *Messenger) handleSyncCommunity(messageState *ReceivedMessageState, sync
+	err = m.communitiesManager.UpdateCommunityDescriptionMagnetlinkMessageClock(community.ID(), magnetLinkMessage.Clock)
+	if err != nil {
+		return err
+	}
```
We're updating the magnet link clock in the community description. This is needed so that, when users receive community invites, they get the latest magnetlink clock (otherwise, the moment they'll join, they'll try to process all past magnetlink messages they receive)

Samyoul commented on protocol/messenger.go:
```diff
@@ -657,6 +660,17 @@ func (m *Messenger) Start() (*MessengerResponse, error) {
 		return nil, err
 	}
 
+	if m.config.torrentConfig.Enabled {
+		adminCommunities, err := m.communitiesManager.Created()
+		if err == nil && len(adminCommunities) > 0 {
+			available := m.SubscribeMailserverAvailable()
+			go func() {
+				<-available
+				m.InitHistoryArchiveTasks(adminCommunities)
+			}()
+		}
+	}
+
```
Ok, I've found the fix for the failing test.

```suggestion
	if m.config.torrentConfig != nil && m.config.torrentConfig.Enabled {
		adminCommunities, err := m.communitiesManager.Created()
		if err == nil && len(adminCommunities) > 0 {
			available := m.SubscribeMailserverAvailable()
			go func() {
				<-available
				m.InitHistoryArchiveTasks(adminCommunities)
			}()
		}
	}

```

0x-r4bbit replied:
Argh... I'm sorry you had to go after this. Thank you!

Samyoul replied:
It's ok, these sorts of bugs are very hard to find. When a package uses `recover` it makes discovery so tricky.

Samyoul commented on protocol/messenger_communities.go:
```diff
@@ -1095,6 +1172,223 @@ func (m *Messenger) handleSyncCommunity(messageState *ReceivedMessageState, sync
+
+			if len(filters) == 0 {
+				log.Println("no filters or chats for this community starting interval", c.IDString())
+				m.logger.Debug("no filters or chats for this community starting interval")
```
I notice a number of double logs like this are these for you own development flow?

Samyoul commented on protocol/protobuf/communities.proto:
```diff
@@ -43,6 +43,7 @@ message CommunityDescription {
   map<string,CommunityChat> chats = 6;
   repeated string ban_list = 7;
   map<string,CommunityCategory> categories = 8;
+  uint64 magnetlink_clock = 9;
```
Just a naming suggestion
```suggestion
  uint64 archive_magnetlink_clock = 9;
```
In case in future communities use magnet links for other things.

Samyoul reviewed (approved):
Really cool PR. This must have taken so much time to do. Very thorough work.

I would replace all usages of `log` with the uber logger library, just as a matter of consistency and so we know the source of our logs.

I also debated with myself about suggesting breaking out the torrent related logic from the `Manager` into a dedicated `TorrentManager`, but I am not sure it is really necessary.

I really like that you chose to use the `waitgroup`, and it is implemented well.

---

Also sorry it took so long to review, there was just a lot going on in this PR.

0x-r4bbit commented:
Removed all usages of `log.Println` and replaced them with `zap.logger`.
Although it was very nice and useful to get the live logs about how the archive protocol behaves. Makes it significantly harder to monitor now.

I've also updated the protobuf property according to you suggestion.

I have not moved the torrent related APIs to a separate manager yet. Happy to do that in the future but I'd prefer landing this PR before introducing another set of bigger changes. Been rebasing this one for a month now :D

ghost commented:
`vendor/github.com/pion/webrtc/v3/yarn.lock` changed. Pinging @adambabik and @corpetty
